{"cells":[{"cell_type":"markdown","metadata":{"formattedRanges":[],"cell_id":"6753c5d2165944e9a2a0431550ddf6ad","deepnote_cell_type":"text-cell-h1"},"source":"# Taxonomy label generation from GTDB-tk"},{"cell_type":"markdown","metadata":{"cell_id":"6e65ad0ca6704f52a7b69c5c7fabb8b2","deepnote_cell_type":"markdown"},"source":"Last updated: June 15 2023\n\nGTDB-tk is a useful tool that classifies bacterial and archaeal genomes using the **Genome Taxonomy Database** (GTDB) nomenclature. Information on how to run GTDB-tk can be found here: https://github.com/Ecogenomics/GTDBTk\n\nUnfortunately, the output classification label is quite long and not practical when referring to a long list of genomes (often the case with metagenome-assembled genomes). This notebook parses the output files and condenses the taxonomy into something more useful for labelling trees, plots, etc. \n\n"},{"cell_type":"markdown","metadata":{"cell_id":"1f3c8284a0994c40b8436162d81074cb","deepnote_cell_type":"markdown"},"source":"### Some examples:\n\nd__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Xanthomonadales;f__Rhodanobacteraceae;g__Metallibacterium;s__Metallibacterium scheffleri --> **Metallibacterium scheffleri**\n\nd__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Acidithiobacillales;f__Acidithiobacillaceae;g__Acidithiobacillus;s__\n--> **Acidithiobacillus sp.**\n\nd__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__UBA2770;f__;g__;s__ --> **c. Gammaproteobacteria (o. UBA2770)**\n\n"},{"cell_type":"markdown","metadata":{"formattedRanges":[],"cell_id":"7959a60f5b98462c85b1f23eaffb1ef8","deepnote_cell_type":"text-cell-p"},"source":"\r\r"},{"cell_type":"markdown","metadata":{"cell_id":"c35ef19e64df491db019e62263a2838a","deepnote_cell_type":"markdown"},"source":"---"},{"cell_type":"code","metadata":{"source_hash":"c76c7c51","execution_start":1686843697105,"execution_millis":46,"deepnote_to_be_reexecuted":false,"cell_id":"3e7218291ad948cb8f56823ce8fc8cac","deepnote_cell_type":"code"},"source":"import pandas as pd\nimport numpy as np","execution_count":1,"outputs":[]},{"cell_type":"markdown","metadata":{"cell_id":"1faa52a9754e4672aa9abbd1e14fac17","deepnote_cell_type":"markdown"},"source":"Importing raw outputs from gtdb-tk: **'gtdbtk.ar53.summary.tsv'** (archaea), and **'gtdbtk.ar53.summary.tsv'** (bacteria)"},{"cell_type":"code","metadata":{"source_hash":"5372893f","execution_start":1686843697149,"execution_millis":2,"deepnote_table_state":{"sortBy":[],"filters":[],"pageSize":10,"pageIndex":0},"deepnote_table_loading":false,"deepnote_to_be_reexecuted":false,"cell_id":"5edff9bec997403f886a4a922757c491","deepnote_cell_type":"code"},"source":"ar53_df = pd.read_table('gtdbtk.ar53.summary.tsv', sep='\\t')\nbac120_df = pd.read_table('gtdbtk.bac120.summary.tsv', sep='\\t')\n\n#filtering only bin ID and taxonomy columns\nar53_df = ar53_df[['user_genome', 'classification']]\nbac120_df = bac120_df[['user_genome', 'classification']]\n\n#concatenating dataframes\ngtdb_all = pd.concat([ar53_df, bac120_df], axis=0)","execution_count":2,"outputs":[]},{"cell_type":"markdown","metadata":{"formattedRanges":[],"cell_id":"2f3df63953e8488ead6d9c24986bfc66","deepnote_cell_type":"text-cell-p"},"source":"Extracting classification data into separate columns and removing the prefixes:"},{"cell_type":"code","metadata":{"source_hash":"cf77bf7a","execution_start":1686843697150,"execution_millis":45,"deepnote_table_state":{"sortBy":[],"filters":[],"pageSize":10,"pageIndex":1},"deepnote_table_loading":false,"deepnote_to_be_reexecuted":false,"cell_id":"28b37ea45cf3445e9fe85a8a259ad0e2","deepnote_cell_type":"code"},"source":"taxonomy_df = gtdb_all.copy()\n\ntaxonomy_df[['domain','phylum', 'class', 'order', 'family', 'genus', 'species']] = \\\ntaxonomy_df['classification'].str.split(';',expand=True)\n\n#removes the first 3 characters from the newly-generated columns:\nfor col in taxonomy_df.columns[2:]:\n    taxonomy_df[col] = taxonomy_df[col].str[3:]\n\n#needed to deal with special cases like 'Leptospirillium_A'\ntaxonomy_df['genus'] = taxonomy_df['genus'].str.replace('_', ' ')\ntaxonomy_df['species'] = taxonomy_df['species'].str.replace('_', ' ')","execution_count":3,"outputs":[]},{"cell_type":"markdown","metadata":{"formattedRanges":[],"cell_id":"23eaa41df17a48fc8a39b6b5c88a9976","deepnote_cell_type":"text-cell-p"},"source":"The following function generates a condensed taxonomy label based on the various levels of classification. It returns the genus and species (or genus + 'sp.'), if known. Otherwise, it takes the lowest level of taxonomy that was assigned a value, and labels it with the appropriate prefix (p., c., o., f.) "},{"cell_type":"code","metadata":{"source_hash":"71fdf2a7","execution_start":1686843697194,"execution_millis":2,"deepnote_table_state":{"sortBy":[],"filters":[],"pageSize":10,"pageIndex":4},"deepnote_table_loading":false,"deepnote_to_be_reexecuted":false,"cell_id":"9d0d1dcf82a14be2bddc9e20d36a8893","deepnote_cell_type":"code"},"source":"def make_taxon_label (df): \n    \n    #sequential conditional values starting from species, check if value exists\n    cond_list = [\n        df['species'] != '',\n        df['genus'] != '',\n        df['family'] != '',\n        df['order'] != '',\n        df['class'] != '',\n        df['phylum'] != ''\n    ]\n\n    #outputs the genus + species, OR the lowest known taxonomy level\n    choice_list = [\n        df['species'],\n        df['genus'] + ' sp.',\n        'f. ' + df['family'],\n        'o. ' + df['order'],\n        'c. ' + df['class'],\n        'p. ' + df['phylum'],\n    ]\n\n    return np.select(cond_list, choice_list) #selects appropriate output based on conditionals","execution_count":4,"outputs":[]},{"cell_type":"markdown","metadata":{"formattedRanges":[],"cell_id":"1de52cc722134a819b9bf8162df7a06e","deepnote_cell_type":"text-cell-p"},"source":"This function filters out any 'non-informative' labels: e.g., 'UBA184' , 'GCA-000496135', so the resulting taxonomy label is more descriptive of the organism's phylogeny at a glance."},{"cell_type":"code","metadata":{"source_hash":"9f5c18a6","execution_start":1686843697194,"execution_millis":13,"deepnote_to_be_reexecuted":false,"cell_id":"2bc826c324a54e4fac7f2a30ac5397a3","deepnote_cell_type":"code"},"source":"#returns an empty string if the string meets any of the following conditions:\n## has numerical digits or special characters (-, _, etc.)\n## is written in all uppercase letters\ndef delete_labels (s):\n    if ( \n        (any(not c.isalnum() for c in s.replace(' ', ''))) or #the replace makes an exception for spaces\n        (any(c.isdigit() for c in s)) or\n        (s.isupper())\n    ):\n        return ''\n    else:\n        return s","execution_count":5,"outputs":[]},{"cell_type":"markdown","metadata":{"formattedRanges":[],"cell_id":"5b600d333f37420586c86a910511fa56","deepnote_cell_type":"text-cell-p"},"source":"Applying functions to the dataframe to generate labels:"},{"cell_type":"code","metadata":{"source_hash":"c7cb194b","execution_start":1686843697214,"execution_millis":71,"deepnote_to_be_reexecuted":false,"cell_id":"20a9f9f0b21b42cb8cb77c68af7099fe","deepnote_cell_type":"code"},"source":"taxon_labels = taxonomy_df.copy()\n\n#initial label that keeps all levels of classification\ntaxon_labels['lab'] = make_taxon_label(taxon_labels)\n\n#new taxonomy df that filters out the non-informative labels\ntaxonomy_filtered = taxonomy_df.copy()\nfor col in taxonomy_filtered.columns[2:]:\n    taxonomy_filtered[col] = taxonomy_filtered[col].apply(delete_labels)\n\n#new set of labels, based on filtered dataset    \ntaxon_labels['filtered_label'] = make_taxon_label(taxonomy_filtered)\n\n#generating the final label by comparing the two label sets:\n#if they match, use original, otherwise concatenate them\ntaxon_labels['label'] = np.where(  \n    taxon_labels['lab'] == taxon_labels['filtered_label'], taxon_labels['lab'],\n    taxon_labels['filtered_label'].astype(str) + ' (' + taxon_labels['lab'].astype(str) + ')')\n\n#adding an additional column which adds the genome ID to the label\ntaxon_labels['bin_w_label'] = taxon_labels['user_genome'] + '; '+ taxon_labels['label']\n\n#filtering and sorting\ntaxon_labels = taxon_labels[['user_genome', 'classification', 'label', 'bin_w_label']]\ntaxon_labels = taxon_labels.sort_values('user_genome', axis=0)\n\ntaxon_labels.to_csv('gtdb_taxon_labels.csv', index_label='user_genome')","execution_count":6,"outputs":[]},{"cell_type":"markdown","source":"<a style='text-decoration:none;line-height:16px;display:flex;color:#5B5B62;padding:10px;justify-content:end;' href='https://deepnote.com?utm_source=created-in-deepnote-cell&projectId=d1ddb408-cd62-40b3-ace3-0d9efd98842d' target=\"_blank\">\n<img alt='Created in deepnote.com' style='display:inline;max-height:16px;margin:0px;margin-right:7.5px;' src='data:image/svg+xml;base64,PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0iVVRGLTgiPz4KPHN2ZyB3aWR0aD0iODBweCIgaGVpZ2h0PSI4MHB4IiB2aWV3Qm94PSIwIDAgODAgODAiIHZlcnNpb249IjEuMSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB4bWxuczp4bGluaz0iaHR0cDovL3d3dy53My5vcmcvMTk5OS94bGluayI+CiAgICA8IS0tIEdlbmVyYXRvcjogU2tldGNoIDU0LjEgKDc2NDkwKSAtIGh0dHBzOi8vc2tldGNoYXBwLmNvbSAtLT4KICAgIDx0aXRsZT5Hcm91cCAzPC90aXRsZT4KICAgIDxkZXNjPkNyZWF0ZWQgd2l0aCBTa2V0Y2guPC9kZXNjPgogICAgPGcgaWQ9IkxhbmRpbmciIHN0cm9rZT0ibm9uZSIgc3Ryb2tlLXdpZHRoPSIxIiBmaWxsPSJub25lIiBmaWxsLXJ1bGU9ImV2ZW5vZGQiPgogICAgICAgIDxnIGlkPSJBcnRib2FyZCIgdHJhbnNmb3JtPSJ0cmFuc2xhdGUoLTEyMzUuMDAwMDAwLCAtNzkuMDAwMDAwKSI+CiAgICAgICAgICAgIDxnIGlkPSJHcm91cC0zIiB0cmFuc2Zvcm09InRyYW5zbGF0ZSgxMjM1LjAwMDAwMCwgNzkuMDAwMDAwKSI+CiAgICAgICAgICAgICAgICA8cG9seWdvbiBpZD0iUGF0aC0yMCIgZmlsbD0iIzAyNjVCNCIgcG9pbnRzPSIyLjM3NjIzNzYyIDgwIDM4LjA0NzY2NjcgODAgNTcuODIxNzgyMiA3My44MDU3NTkyIDU3LjgyMTc4MjIgMzIuNzU5MjczOSAzOS4xNDAyMjc4IDMxLjY4MzE2ODMiPjwvcG9seWdvbj4KICAgICAgICAgICAgICAgIDxwYXRoIGQ9Ik0zNS4wMDc3MTgsODAgQzQyLjkwNjIwMDcsNzYuNDU0OTM1OCA0Ny41NjQ5MTY3LDcxLjU0MjI2NzEgNDguOTgzODY2LDY1LjI2MTk5MzkgQzUxLjExMjI4OTksNTUuODQxNTg0MiA0MS42NzcxNzk1LDQ5LjIxMjIyODQgMjUuNjIzOTg0Niw0OS4yMTIyMjg0IEMyNS40ODQ5Mjg5LDQ5LjEyNjg0NDggMjkuODI2MTI5Niw0My4yODM4MjQ4IDM4LjY0NzU4NjksMzEuNjgzMTY4MyBMNzIuODcxMjg3MSwzMi41NTQ0MjUgTDY1LjI4MDk3Myw2Ny42NzYzNDIxIEw1MS4xMTIyODk5LDc3LjM3NjE0NCBMMzUuMDA3NzE4LDgwIFoiIGlkPSJQYXRoLTIyIiBmaWxsPSIjMDAyODY4Ij48L3BhdGg+CiAgICAgICAgICAgICAgICA8cGF0aCBkPSJNMCwzNy43MzA0NDA1IEwyNy4xMTQ1MzcsMC4yNTcxMTE0MzYgQzYyLjM3MTUxMjMsLTEuOTkwNzE3MDEgODAsMTAuNTAwMzkyNyA4MCwzNy43MzA0NDA1IEM4MCw2NC45NjA0ODgyIDY0Ljc3NjUwMzgsNzkuMDUwMzQxNCAzNC4zMjk1MTEzLDgwIEM0Ny4wNTUzNDg5LDc3LjU2NzA4MDggNTMuNDE4MjY3Nyw3MC4zMTM2MTAzIDUzLjQxODI2NzcsNTguMjM5NTg4NSBDNTMuNDE4MjY3Nyw0MC4xMjg1NTU3IDM2LjMwMzk1NDQsMzcuNzMwNDQwNSAyNS4yMjc0MTcsMzcuNzMwNDQwNSBDMTcuODQzMDU4NiwzNy43MzA0NDA1IDkuNDMzOTE5NjYsMzcuNzMwNDQwNSAwLDM3LjczMDQ0MDUgWiIgaWQ9IlBhdGgtMTkiIGZpbGw9IiMzNzkzRUYiPjwvcGF0aD4KICAgICAgICAgICAgPC9nPgogICAgICAgIDwvZz4KICAgIDwvZz4KPC9zdmc+' > </img>\nCreated in <span style='font-weight:600;margin-left:4px;'>Deepnote</span></a>","metadata":{"created_in_deepnote_cell":true,"deepnote_cell_type":"markdown"}}],"nbformat":4,"nbformat_minor":0,"metadata":{"deepnote":{},"orig_nbformat":2,"deepnote_notebook_id":"696165cb86504db98e1eaef5933fa7b6","deepnote_persisted_session":{"createdAt":"2023-06-13T19:16:52.120Z"},"deepnote_execution_queue":[]}}